pkg/dataobj/planner/logical/ifc.go

```diff
@@ -0,0 +1,201 @@
+// Package logical implements logical query plan operations and expressions
+package logical
+
+import (
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/datasetmd"
+	"github.com/grafana/loki/v3/pkg/dataobj/planner/schema"
+)
+
+/*
+Interface Design Pattern for Query Plans and Expressions
+
+This package uses a type-based polymorphism pattern that allows for:
+
+1. Extension - Adding new concrete implementations without modifying consumer code
+2. Type Safety - Compile-time guarantees through interfaces and type assertions
+3. API Surface Reduction - Exposing minimal interfaces to external consumers
+
+The pattern works as follows:
```

Comment:
Not sure about this level of indirection. WDYT

Reply:
Don't have a strong opinion. But I think it's not strictly necessary to have a `Type()` function, since you can use the type switch statement instead.